cpp/src/arrow/util/future.h

```diff
@@ -0,0 +1,318 @@
+
+  friend class FutureImpl;
+  friend class ConcreteFutureImpl;
+};
+
+// ---------------------------------------------------------------------
+// An intermediate class for storing Future results
+
+class FutureStorageBase {
+ public:
+  FutureStorageBase() : impl_(FutureImpl::Make()) {}
+
+ protected:
+  ARROW_DISALLOW_COPY_AND_ASSIGN(FutureStorageBase);
+  std::unique_ptr<FutureImpl> impl_;
+
+  template <typename T>
+  friend class Future;
+};
+
+template <typename T>
+class FutureStorage : public FutureStorageBase {
+ public:
+  Status status() const { return result_.status(); }
+
+  template <typename U>
+  void MarkFinished(U&& value) {
+    result_ = std::forward<U>(value);
+    if (ARROW_PREDICT_TRUE(result_.ok())) {
+      impl_->MarkFinished();
+    } else {
+      impl_->MarkFailed();
+    }
+  }
+
+  template <typename Func>
+  void ExecuteAndMarkFinished(Func&& func) {
+    MarkFinished(func());
+  }
+
+ protected:
+  Result<T> result_;
+  friend class Future<T>;
+};
+
+// A Future<void> just stores a Status (always ok for now, but that could change
+// if we implement cancellation).
+template <>
+class FutureStorage<void> : public FutureStorageBase {
+ public:
+  Status status() const { return status_; }
+
+  void MarkFinished(Status st = Status::OK()) {
+    status_ = std::move(st);
+    impl_->MarkFinished();
+  }
+
+  template <typename Func>
+  void ExecuteAndMarkFinished(Func&& func) {
+    func();
+    MarkFinished();
+  }
+
+ protected:
+  Status status_;
+};
+
+// A Future<Status> just stores a Status.
+template <>
+class FutureStorage<Status> : public FutureStorageBase {
+ public:
+  Status status() const { return status_; }
+
+  void MarkFinished(Status st) {
+    status_ = std::move(st);
+    if (ARROW_PREDICT_TRUE(status_.ok())) {
+      impl_->MarkFinished();
+    } else {
+      impl_->MarkFailed();
+    }
+  }
+
+  template <typename Func>
+  void ExecuteAndMarkFinished(Func&& func) {
+    MarkFinished(func());
+  }
+
+ protected:
+  Status status_;
+};
+
+// ---------------------------------------------------------------------
+// Public API
+
+template <typename T>
+class Future {
+ public:
+  static constexpr double kInfinity = FutureImpl::kInfinity;
+
+  // Consumer API
+
+  FutureState state() const { return impl_->state(); }
+
+  const Result<T>& result() const& {
+    Wait();
+    return storage_->result_;
+  }
+
+  Result<T> result() && {
+    Wait();
+    return std::move(storage_->result_);
+  }
+
+  Status status() const {
+    Wait();
+    return storage_->status();
+  }
+
+  void Wait() const {
+    if (!IsFutureFinished(impl_->state())) {
+      impl_->Wait();
+    }
+  }
+
+  bool Wait(double seconds) const {
+    if (IsFutureFinished(impl_->state())) {
+      return true;
+    }
+    return impl_->Wait(seconds);
+  }
+
+  // Producer API
+
+  template <typename Func>
+  void ExecuteAndMarkFinished(Func&& func) {
+    storage_->ExecuteAndMarkFinished(std::forward<Func>(func));
```

Comment:
This doesn't need to be a member of storage_

Reply:
Yeah, well, I spent enough time trying to find a non-fucked up way of spelling out the different specializations for `void` and `Status` that I don't want to go through this again, sorry.